contrib/build.sh

```diff
@@ -0,0 +1,40 @@
+#! /bin/bash
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+VERSION="1"
+IMAGE="accumulo-build-environment-${VERSION}"
```

Comment:
My understanding is: this image doesn't exist. We'd have to create this?

Reply:
If it doesn't exist on the machine, then the script creates it further down in the script.

Reply:
Could this not be built and published to GH packages? I guess in that case you couldn't set the uid/gid to match, but you could still build/publish a base image. Then have the script extend that base image and only define a user matching the uid/gid of the local user. Having a published base image would allow you to use the image for things like GH actions and/or codespaces. It would also speed up the local build for most people since they'd only have to download the base image vs wait for thrift to build while building the image. The base image itself could be built and published using GH actions.

Reply:
I just thought of that, I think there might be a way so specify the uid at runtime. I'll try it out.

Reply:
You can definitely use the `-u`/`--user` arg and specify a `uid:gid` pair. I'm not sure what happens if you specify a uid and/or gid that doesn't exist in the image. I _think_ it will work, but am not positive.

Reply:
I about to find out....

Reply:
The `-u` argument at runtime is not working. I'll push up what I have if you want to mess around with it.

Reply:
The `-u` argument works, but since the user id doesn't exist in /etc/passwd, there's no home dir for the user. I was able to work around that by making a world writable dir in the Docker image and adding env vars HOME (pointing to the new dir) and MAVEN_OPTS (setting user.home to the dir). That gets the build to work if you don't run tests, but tests fail on setting up miniaccumulo. There are other workarounds where you could specify an entrypoint script that modifies the user, but I think it's probably better just to create a new image layered on top of the base image. Something like this for the user-specific Dockerfile:
```Dockerfile
ARG DOCKER_PROXY=""
ARG version=1
FROM ghcr.io/apache/accumulo/accumulo-build-environment:${version}
ARG uid
ARG gid

USER root

# Update uid/gid of the builder user to the supplied ids. Handle conflicts by removing the conflicting user or group.
RUN bash -c "groupmod -g ${gid} builder || (groupdel -f $(getent group ${gid} | cut -d: -f1) && groupmod -g ${gid} builder)" && \
    bash -c "usermod -u ${uid} builder || (userdel $(id -nu ${uid}) && usermod -u ${uid} builder)"

USER builder
```
Not sure about the path to the base image--would be something like that depending on how it gets published.

Reply:
@brianloss - Looking at [this](https://www.fullstaq.com/knowledge-hub/blogs/docker-and-the-host-filesystem-owner-matching-problem), which describes the problem pretty well, I was able to modify the Docker image to be generic and modify the `builder` user at runtime in 7f93364.